matrix/matrix.py

```diff
@@ -0,0 +1,121 @@
+from typing import Union
+
+
+class Matrix:
+
+    def __init__(self, matrix: tuple):
+        self.is_valid_matrix(matrix)
+        self.matrix = matrix
+
+    @property
+    def tuples(self):
+        return self.matrix
+
+    def __str__(self):
+        return str(self.matrix)
+
+    def __repr__(self):
+        return f"Matrix({self.matrix})"
+
+    def __len__(self):
+        return len(self.matrix)
+
+    def __getitem__(self, i):
+        return self.matrix[i]
+
+    def __hash__(self):
+        return hash(self.matrix)
+
+    def __neg__(self):
+        return self * (-1)
+
+    def __eq__(self, other):
+        if type(other) is not Matrix:
+            raise TypeError("Argument is not of Matrix type.")
+
+        if len(self) != len(other):
+            return False
+
+        if any(s_row != o_row for s_row, o_row in zip(self, other)):
+            return False
+
+        return True
+
+    def __ne__(self, other):
+        return not self == other
+
+    def __add__(self, other):
+        if type(other) is not Matrix:
+            raise TypeError("Argument is not of Matrix type.")
+
+        if len(self) != len(other):
+            raise ValueError("Matrix must be of the same order.")
+
+        return Matrix(tuple(tuple((sum(cell_pair) for cell_pair in zip(self_row, other_row)))
+                            for self_row, other_row in zip(self, other)))
+
+    def __sub__(self, other):
+        if type(other) is not Matrix:
+            raise TypeError("Argument is not of Matrix type.")
+
+        return self + (-other)
+
+    def __mul__(self, other):
+        if type(other) is int or type(other) is float:
```

Comment:
I still suggest using a decorator to specify the types that are eligible for this function.
then you can remove all the type checking inside the function and keep only the logic.